File: MauloaDemo.Models/Church.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using CBAF;

namespace MauloaDemo.Models {

    [Table("church")]
    public class Church : IValidatableObject {

        [Key(), Required(), StringLength(5)]
        public string church_cd { get; set; }

        [Required, StringLength(100)]
        public string church_name { get; set; }

        [Required, StringLength(100)]
        public string church_name_jpn { get; set; }

        [StringLength(1)]
        public string plan_kind { get; set; }

        [StringLength(10)]
        public string abbrev_jpn { get; set; }

        [StringLength(10)]
        public string abbrev_eng { get; set; }

        [StringLength(3)]
        public string area_cd { get; set; }

        public bool exclusive { get; set; }

        public short? release_days { get; set; }

        [StringLength(100)]
        public string op_address1 { get; set; }

        [StringLength(100)]
        public string op_address2 { get; set; }

        [StringLength(100)]
        public string op_address3 { get; set; }

        [StringLength(100)]
        public string op_address4 { get; set; }

        [StringLength(100)]
        public string op_tel { get; set; }

        [StringLength(100)]
        public string op_fax { get; set; }

        [StringLength(100)]
        public string ac_address1 { get; set; }

        [StringLength(100)]
        public string ac_address2 { get; set; }

        [StringLength(100)]
        public string ac_address3 { get; set; }

        [StringLength(100)]
        public string ac_address4 { get; set; }

        [StringLength(100)]
        public string ac_tel { get; set; }

        [StringLength(100)]
        public string ac_fax { get; set; }

        public string note { get; set; }

        public DateTime? discon_date { get; set; }

        public short disp_seq { get; set; }

        public short default_pickup { get; set; }

        [Required]
        public string last_person { get; set; }

        [CBAF.Attributes.IgnoreChangeDiff]
        public DateTime update_date { get; set; }




        [NotMapped]
        public bool is_new { get; set; }

        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext) {
            if (!string.IsNullOrEmpty(this.church_cd) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.church_cd, false)) {
                yield return new ValidationResult(string.Format("church_cd は半角英数字で入力してください。({0})", this.church_cd), new[] { "church_cd" });
            }

            if (!string.IsNullOrEmpty(this.church_name) && !HankakuZenkaku.isHankakuAlphaNumericOrSymbol(this.church_name, true)) {
                yield return new ValidationResult(string.Format("Church Name Eng は半角英数字で入力してください。({0})", this.church_name), new[] { "church_name" });
            }
        }

    }
}

File: MauloaDemo.Web/Areas/Masters/Scripts/Models/Church.ts
//TypeScript用型定義
interface IChurch extends kendo.data.Model {
    //インスタンスメソッド・プロパティ
    parseJSON(): void;
    save(): JQueryXHR;
    destroy(): JQueryXHR;
}
interface ChurchFn {
    //スタティックメソッド・プロパティ
    new (options?: any): IChurch;
    prototype: IChurch;
    fetch(church_cd: string, callback: JQueryPromiseCallback<IChurch> ): JQueryDeferred<IChurch>;
    getDataSource(data? : any): kendo.data.DataSource;
}
//クラス名を定義
declare var Church: ChurchFn;


define(function () {

    var Church: ChurchFn = <any>kendo.data.Model.define({
        id: "church_cd",
        fields: {
            "church_cd": { type: "string", validation: { required: true } },
            "church_name": { type: "string" },
            "church_name_jpn": { type: "string" },
            "plan_kind": { type: "string" },
            "abbrev_jpn": { type: "string" },
            "abbrev_eng": { type: "string" },
            "area_cd": { type: "string" },
            "exclusive": { type: "string" },
            "release_days": { type: "string" },
            "op_address1": { type: "string" },
            "op_address2": { type: "string" },
            "op_address3": { type: "string" },
            "op_address4": { type: "string" },
            "op_tel": { type: "string" },
            "op_fax": { type: "string" },
            "ac_address1": { type: "string" },
            "ac_address2": { type: "string" },
            "ac_address3": { type: "string" },
            "ac_address4": { type: "string" },
            "ac_tel": { type: "string" },
            "ac_fax": { type: "string" },
            "note": { type: "string" },
            "discon_date": { type: "date", nullable: true, defaultValue: null },
            "disp_seq": { type: "string" },
            "default_pickup": { type: "string" },
            "last_person": { type: "string" },
            "update_date": { type: "date", defaultValue: null }
        }
    });

    Church.prototype.toJSON = function(){
        var json = kendo.data.ObservableObject.prototype.toJSON.call(this); //Call the original toJSON method
        delete json.dirty;
        delete json.id;
        json.discon_date = kendo.toString(this.discon_date, "yyyy/MM/dd");
        json.update_date = kendo.toString(this.update_date, "yyyy/MM/dd HH:mm:ss");
        return json;
    };

    Church.prototype.parseJSON = function(){
        this.set("id", this.get(this.idField)); //idを明示的にセット。isNew()の値を正しく取得するために必要。
        this.discon_date = App.Utils.convertToDateTime(this.discon_date);
        this.update_date = App.Utils.convertToDateTime(this.update_date);
        this.dirty = false;
    };


    Church.prototype.save = function(){
        this.set("is_new", this.isNew());
        var dataStr = JSON.stringify(this);

        var deffered = $.ajax({
            url: App.getApiPath("Churches/Save"),
            type: "POST",
            data: dataStr,
            processData: false,
            contentType: "application/json; charset=utf-8"
        });
        return deffered;
    };

    Church.prototype.destroy = function(){
        var deffered = $.ajax({
            url: App.getApiPath("Churches/Delete/" + encodeURIComponent(this.get("church_cd"))),
            type: "POST",
            processData: false,
            contentType: "application/json; charset=utf-8"
        });
        return deffered;
    };

    
    Church.fetch = function (church_cd, callback: JQueryPromiseCallback<IChurch>) : JQueryDeferred<IChurch> {
        var deffered = $.ajax({
            url: App.getApiPath("Churches/" + encodeURIComponent(church_cd)),
                type: "GET",
                cache: false
            })
            .done(function (data) {
                var model = $.extend(new Church(), data);
                model.parseJSON();
                if (callback) callback(model);
            });
        return deffered;
    };

    Church.getDataSource = function (data) : kendo.data.DataSource {
        return new kendo.data.DataSource({
            transport: {
                read: {
                    url: App.getApiPath("Churches/Search"),
                    type: "POST",
                    dataType: "json",
                    data: data
                }
            },
            schema: { model: Church }
        });
    };

    return Church;
});

